lib/ocpp/v201/functional_blocks/data_transfer.cpp

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Pionix GmbH and Contributors to EVerest
+
+#include <ocpp/common/constants.hpp>
+#include <ocpp/v201/functional_blocks/data_transfer.hpp>
+
+namespace ocpp {
+namespace v201 {
+
+void DataTransfer::handle_data_transfer_req(Call<DataTransferRequest> call) {
+    const auto msg = call.msg;
+    DataTransferResponse response;
+    response.status = DataTransferStatusEnum::UnknownVendorId;
+
+    if (this->data_transfer_callback.has_value()) {
+        response = this->data_transfer_callback.value()(call.msg);
+    } else {
+        response.status = DataTransferStatusEnum::UnknownVendorId;
+        EVLOG_warning << "Received a DataTransferRequest but no data transfer callback was registered";
+    }
+
+    ocpp::CallResult<DataTransferResponse> call_result(response, call.uniqueId);
+    this->message_dispatcher.dispatch_call_result(call_result);
+}
+
+std::optional<DataTransferResponse> DataTransfer::data_transfer_req(const CiString<255>& vendorId,
+                                                                    const std::optional<CiString<50>>& messageId,
+                                                                    const std::optional<json>& data) {
+    DataTransferRequest req;
+    req.vendorId = vendorId;
+    req.messageId = messageId;
+    req.data = data;
+
+    return this->data_transfer_req(req);
+}
+
+std::optional<DataTransferResponse> DataTransfer::data_transfer_req(const DataTransferRequest& request) {
+    DataTransferResponse response;
+    response.status = DataTransferStatusEnum::Rejected;
+
+    ocpp::Call<DataTransferRequest> call(request);
+    auto data_transfer_future = this->message_dispatcher.dispatch_call_async(call);
+
+    if (not this->is_websocket_connected()) {
```

Comment:
I don't really like that we have to pass this check along here. Especially since we also have `enhanced_message.offline` as a value that we return. Would be nice if we can remove the is connected callback and just use the other behavior instead.

If we really do need to have information if we are connected, maybe its better to abstract it into the message dispatcher instead since all blocks will have that anyway.

Reply:
I agree, I was just trying to move the functionality to the functional block and provide everything that is required. I think we can indeed just make use of `enhanced_message.offline` here.